### PR TITLE
Ensure UI theme context is available in test renders

### PR DESCRIPTION
**Pull request by kofi-q**
## Overview
Related to: https://github.com/votingworks/vxsuite/issues/2744

- Adding a drop-in replacement for the `render` function from `@testing-library/react` that wraps test content in the `libs/ui/AppBase`  component to provide global styles and theme context needed by components updated to use themes.
- Following the recommendation from the [testing library docs](https://testing-library.com/docs/react-testing-library/setup/#custom-render) to re-export `@testing-library/react` with the `render` function overridden.
- Just did a replace-all `@testing-library/react` -> `../test/react_testing_library` for imports in the codebase - really only needed for any `render` imports, but it was simpler to just replace all imports, to avoid confusion.
- Would've liked to put the re-export code in `libs/test-utils`, but it would create a circular dependency with libs/ui

## Demo Video or Screenshot
N/A

## Testing Plan
- Added a test for the `renderWithThemes` utility to verify that using the plain `render` fails on themed components, but works fine with `renderWithThemes`
- Updated changed jest snapshots as needed

## Checklist

- [n/a] I have added
      [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging)
      where appropriate to any new user actions, system updates such as file
      reads or storage writes, or errors introduced.
- [x] I have added JSDoc comments to any newly introduced exports

- [n/a] I have added a screenshot and/or video to this PR to demo the change
- [n/a] I have added the "user_facing_change" label to this PR to automate an
      announcement in #machine-product-updates

### CODE REVIEW

File: apps/admin/frontend/src/app.test.tsx [hunk 1]
```diff
@@ -318,11 +312,18 @@ test('L&A (logic and accuracy) flow', async () => {
...
-  renderRootElement(<App hardware={hardware} printer={printer} />, {
-    apiClient: mockApiClient,
-    backend,
-    logger,
-  });
+  renderRootElement(
+    <App
+      hardware={hardware}
+      printer={printer}
+      generateBallotId={mockRandomBallotId}
```
[the remark this answers is not in the document]
Comment (adghayes) [reply apps/admin/frontend/src/app.test.tsx:319]: I tried mocking `votingworks/utils` in a branch and it seemed to work: https://github.com/votingworks/vxsuite/commit/e688b3f275fed8a163d1c7afe482dcb72bcb25bb. Maybe it had something to do with your environment or something you were doing earlier? Or maybe it only applies to VxMark and not VxAdmin, as I didn't try VxMark. Or maybe I'm missing something.

I'd prefer if we avoid doing this. It seems like a non-trivial change for testing purposes and it seems strange to pass in `generateBallotId` at this level.

**Review by adghayes — COMMENTED** [review]
This may be a dumb question, but is there any reason why we couldn't "re-export" the updated `testing-library` from `libs/ui` itself?

I think it makes sense to group test utils with their related code rather than separately (e.g. in `libs/ui` rather than `libs/test-utils`), like @jonahkagan did for `grout/test-utils`. But ultimately I don't think the way we re-export is very important just a nice-to-have if it were consolidated.

Mostly, I want to see if it's really necessary to inject the `generateBallotId`.

Everything else LGTM.

File: apps/admin/frontend/src/app.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kofi-q) [reply apps/admin/frontend/src/app.test.tsx:319]: Ah, thanks for trying that out! It's going to be frustrating if it turns out to be a weird env thing, but also a relief 😅 - I'll give it another shot

**Review by kofi-q — COMMENTED** [review]
Didn't realise we could nest sub-packages! I initially tried doing the re-exports in `libs/ui/testing-library` and couldn't figure out how to get it working and hit some eslint errors as well - https://votingworks.slack.com/archives/CEL6D3GAD/p1676491020665519

I think the grout example solves all my woes - thanks!

File: apps/admin/frontend/src/app.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kofi-q) [reply apps/admin/frontend/src/app.test.tsx:319]: Ok, your suspicion was right - the VxMark tests started breaking again when I reverted to using the jest module mock, so I'm just going to leave this as it is, for now.
In any case, I feel using jest module mocks is usually a sign that the code wasn't designed with testability in mind, so the fewer cases of that, the better

File: libs/ui/src/themes/render_with_themes.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+export function renderWithThemes(
+  ui: React.ReactElement,
+  options: VxRenderOptions = {}
+): RenderResult {
+  const { vxTheme = {}, ...passthroughOptions } = options;
+
+  function wrapper(props: { children: React.ReactNode }): JSX.Element {
+    return (
+      <AppBase
+        colorMode={vxTheme.colorMode}
```
Comment (kofi-q) [inline libs/ui/src/themes/render_with_themes.tsx:28]: Just added support for specifying theme params while rendering to make life a little easier in future PRs

**Review by kofi-q — COMMENTED** [review]
Ok, I looked into doing the re-exports in a nested package, but ran into the same circular dependency issue, since that package will need to depend on `AppBase` from libs/ui and libs/ui would need to depend on the new package for tests.

LMK if you think we should simplify by only exporting `renderWithThemes` - didn't make any changes to avoid thrashing too much, but hoping there's a path to future consolidation involving adding an exception to the eslint rule that stops us from doing something along the lines of `import {render} from '@votingworks/ui/src/testing-library'`

**Review by adghayes — APPROVED** [review]
@kofi-q I agree with the current approach of exporting the whole package, as we might be overriding other methods for RTL and it's the recommendation.

I am thinking you could export the package, in a namespace, from `libs/ui` and then have a package whose only purpose is to import and export that same package, not namespace-d. I think right now the re-exporting code is only repeated in each of the frontends and we have only 4, maybe 5 soon, so I don't mind just leaving it as-is. 

I've never felt there was a problem using `jest` to mock modules, but see where you're coming from. It's a different mindset than I've had, focusing on code being testable rather than code being production-centric and testing around that. 

The VxMark issue concerns me because we don't understand why it's happening. The other changes in this PR don't connect clearly to it. And I am guessing the mocking of other exports from `@votingworks/utils` won't work in VxMark either, which could create problems later.

In any case, everything else in the PR looks good and I trust your troubleshooting.

**Pull request merged**